.github/workflows/push-mimir-build-image.yml

```diff
@@ -0,0 +1,141 @@
+      - name: Get Variables
+        id: variables
+        run: |
+          # The case when workflow is triggered by pull request
+          if [[ ${{ github.event_name }} == 'pull_request' && ${{ github.event.pull_request.user.login }} == 'dependabot[bot]' ]]; then
+            echo "triggered=true" >> $GITHUB_OUTPUT
+            echo "pull_request_number=${{ github.event.pull_request.number }}" >> $GITHUB_OUTPUT
+            echo "pull_request_url=${{ github.event.pull_request.html_url }}" >> $GITHUB_OUTPUT
+            echo "login=${{ github.event.pull_request.user.login }}" >> $GITHUB_OUTPUT
+          # The case when workflow is triggered by comment on pull request and it matches the trigger
+          elif [[ ${{ github.event_name }} == 'issue_comment' && ${{ steps.trigger.outputs.triggered }} == 'true' ]]; then
+            echo "triggered=true" >> $GITHUB_OUTPUT
+            echo "pull_request_number=${{ github.event.issue.number }}" >> $GITHUB_OUTPUT
+            # echo "pull_request_url=${{ github.event.issue.html_url }}" >> $GITHUB_OUTPUT
+            echo "login=${{ github.event.issue.author }}" >> $GITHUB_OUTPUT
+          else
+            echo "triggered=false" >> $GITHUB_OUTPUT
+          fi
+    outputs:
+      triggered: ${{ steps.variables.outputs.triggered }}
+      pull_request_number: ${{ steps.variables.outputs.pull_request_number }}
+      pull_request_url: ${{ steps.variables.outputs.pull_request_url }}
+      login: ${{ steps.variables.outputs.login }}
+
+  build_and_push:
+    runs-on: ubuntu-latest
+    needs: prepare_variables
+    if: needs.prepare_variables.outputs.triggered == 'true'
+    # If the PR is raised by dependabot, then build the image
+    steps:
+      - name: Checkout Repository
+        uses: actions/checkout@v3
+
+      - name: Checkout Pull Request Branch
+        run: gh pr checkout ${{ needs.prepare_variables.outputs.pull_request_number }}
+        env:
+          GITHUB_TOKEN: ${{secrets.GITHUB_TOKEN}}
+
+      - name: Setup QEMU
+        uses: docker/setup-qemu-action@v2
+
+      - name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v2
+
+      - name: Login to Docker Hub
+        uses: docker/login-action@v2
+        with:
+          username: ${{ secrets.DOCKER_USERNAME }}
+          password: ${{ secrets.DOCKER_PASSWORD }}
+
+      - name: Prepare Variables
+        id: prepare
+        run: |
+          echo "path=mimir-build-image/Dockerfile" >> $GITHUB_OUTPUT
+          echo "image=grafana/mimir-build-image" >> $GITHUB_OUTPUT
+
+      - name: Compute Image Tag
+        id: compute_hash
+        run: |
+          current_hash=$(md5sum ${{ steps.prepare.outputs.path }} | awk '{print $1}')
+          echo "the file path is ${{ steps.prepare.outputs.path }}"
+          echo "build tag is $current_hash"
+          tag="${{ needs.prepare_variables.outputs.pull_request_number }}_$current_hash"
+          echo "tag=$tag" >> $GITHUB_OUTPUT
```

Comment:
If we are going to compute the image tag by taking a checksum of the build image Dockerfile, should we just do this in the Makefile when determining the image tag? If we do it in the Makefile, then it removes the need to modify the Makefile every time the build image changes.

Reply:
What you describe was the original intent, but there's a problem... if PR modifies build image Dockerfile, how will Makefile find *previous* build image to actually run actions with? New build image doesn't exist yet, so it can't be used.

Reply:
If we do these steps (checking that the build image has been built and rebuilding it if it hasn't) at the start of the `prepare` job in the existing CI workflow, then all later steps can use the new image. 

Then, the only steps that need to run without the build image are computing the build image tag (which has to run outside the build image always anyway), and checking if the build image exists.

Reply:
👍

Reply:
We were hoping to include additional information in the image tag: PR number that last changed the docker-image (to provide some trace where the Dockerfile update came from), and Go version included in the image. When computing the tag, including Go version would be easy (grep dockerfile), but PR number would be tricky.

Reply:
One more scenario came up in discussion: what if there are two PRs updating build-image at the same time. When computing the tag, they both could get merged, and then final integration of the two changes would happen on the `main` branch -- when running "prepare" job again.

If we keep reference to the tag instead, then after merging one PR, second would have a conflict and merge would be blocked until conflict is resolved. That would move integration of the two changes into the PR.

(In practice we don't update build image that often.)